devdoc/sealable_channel.md

```diff
@@ -0,0 +1,258 @@
+
+If `current_channel_item_count` + 1 is greater than `max_channel_item_count`, `sealable_channel_push` shall fail and return `SEALABLE_CHANNEL_RESULT_OVER_CAPACITY`.
+
+`sealable_channel_push` shall add `data->size` to `current_channel_item_size`.
+
+`sealable_channel_push` shall increment the `current_channel_item_count`.
+
+`sealable_channel_push` shall call `critical_section_leave`.
+
+`sealable_channel_push` shall allocate an `on_underlying_data_consumed_context".
+
+`sealable_channel_push` shall call `channel_push`, passing `on_underling_data_consumed_callback`, and `on_underlying_data_consumed_context`
+
+`sealable_channel_push` shall call `sm_exec_end`.
+
+`sealable_channel_push` shall succeed and return `SEALABLE_CHANNEL_RESULT_OK`.
+
+If any error occurs, `sealable_channel_push` shall undo changes to `current_channel_item_size` and `current_channel_item_count`.
+
+If any error occurs, `sealable_channel_push` shall fail and return `SEALABLE_CHANNEL_RESULT_ERROR`.
+
+### on_underlying_data_consumed
+
+```c
+static void on_underlying_data_consumed(void* push_context, CHANNEL_CALLBACK_RESULT result, THANDLE(RC_STRING) pull_correlation_id, THANDLE(RC_STRING) push_correlation_id);
+```
+
+``on_underlying_data_consumed`` is called by the underlying `channel` when data is pulled from channel.
+It exists primary to translate callback enums from `CHANNEL_CALLBACK_RESULT` to `SEALABLE_CHANNEL_CALLBACK_RESULT` values.
+
+If `push_context` is NULL, `on_underlying_data_consumed` shall terminate the process.
+
+`on_underlying_data_consumed` shall call the original `on_data_consumed_cb`.
+
+`on_underlying_data_consumed` shall free the `on_underlying_data_consumed_context`.
+
+### sealable_channel_get_reporting_info
+
+```c
+int sealable_channel_get_reporting_info(THANDLE(SEALABLE_CHANNEL) sealable_channel, SEALABLE_CHANNEL_REPORTING_INFO* reporting_info);
+```
+
+`sealable_channel_get_reporting_info` returns reporting information for the `sealable_channel`.
+
+If `sealable_channel` is `NULL`, `sealable_channel_get_reporting_info` shall fail and return a non-zero value.
+
+If `reporting_info` is `NULL`, `sealable_channel_get_reporting_info` shall fail and return a non-zero value.
+
+`sealable_channel_get_reporting_info` shall copy information into the `reporting_info` structure.
+
+`sealable_channel_get_reporting_info` shall succeed and return 0.
+
+### sealable_channel_seal_channel
+
+```c
+int sealable_channel_seal_channel(THANDLE(SEALABLE_CHANNEL) sealable_channel);
+```
+
+`sealable_channel_seal_channel` seals the `sealable_channel`.
+
+If `sealable_channel` is `NULL`, `sealable_channel_seal_channel` shall fail and return a non-zero value.
+
+`sealable_channel_seal_channel` shall call `sm_exec_begin`
+
+`sealable_channel_seal_channel` shall mark the `sealable_channel` as sealed.
```

Comment:
What happens (or what should happen) if you seal a channel, close it, and open it. Is it sealed or not?